models/airswap/ethereum/airswap_ethereum_trades.sql

```diff
@@ -1,186 +1,146 @@
+        cast(NULL as double) AS amount_usd,
+        evt_index
+    FROM {{ source('airswap_ethereum', 'Swap_v3_evt_Swap')}} e
+    {% if is_incremental() %}
+    WHERE evt_block_time >= date_trunc("day", now() - interval '1 week')
+    {% endif %}
 )
-
 SELECT
-    'ethereum' AS blockchain,
-    project,
-    version,
-    TRY_CAST(date_trunc('DAY', dexs.block_time) AS date) AS block_date,
-    dexs.block_time,
-    token_a_symbol AS token_bought_symbol,
-    token_b_symbol AS token_sold_symbol,
-    case
-        when lower(token_a_symbol) > lower(token_b_symbol) then concat(token_b_symbol, '-', token_a_symbol)
-        else concat(token_a_symbol, '-', token_b_symbol)
-    end AS token_pair,
-    token_a_amount AS token_bought_amount,
-    token_b_amount AS token_sold_amount,
-    token_a_amount_raw AS token_bought_amount_raw,
-    token_b_amount_raw AS token_sold_amount_raw,
-    usd_amount AS amount_usd,
-    token_a_address AS token_bought_address,
-    token_a_address AS token_sold_address,
-    trader_a AS taker,
-    trader_b AS maker,
-    exchange_contract_address AS project_contract_address,
-    tx_hash,
-    tx_from,
-    tx_to,
-    trace_address,
-    evt_index,
-    trade_id
-FROM dexs
\ No newline at end of file
+    'ethereum' AS blockchain
+    ,'airswap' AS project
+    ,version
+    ,TRY_CAST(date_trunc('DAY', dexs.block_time) AS date) AS block_date
+    ,dexs.block_time
+    ,erc20a.symbol AS token_bought_symbol
+    ,erc20b.symbol AS token_sold_symbol
+    ,case
+        when lower(erc20a.symbol) > lower(erc20b.symbol) then concat(erc20b.symbol, '-', erc20a.symbol)
+        else concat(erc20a.symbol, '-', erc20b.symbol)
+    end as token_pair
+    ,dexs.token_bought_amount_raw / power(10, erc20a.decimals) AS token_bought_amount
+    ,dexs.token_sold_amount_raw / power(10, erc20b.decimals) AS token_sold_amount
+    ,dexs.token_bought_amount_raw
+    ,dexs.token_sold_amount_raw
+    ,coalesce(
+        dexs.amount_usd
+        ,(dexs.token_bought_amount_raw / power(10, p_bought.decimals)) * p_bought.price
+        ,(dexs.token_sold_amount_raw / power(10, p_sold.decimals)) * p_sold.price
+    ) AS amount_usd
+    ,dexs.token_bought_address
+    ,dexs.token_sold_address
+    ,coalesce(dexs.taker, tx.from) AS taker -- subqueries rely on this COALESCE to avoid redundant joins with the transactions table
+    ,dexs.maker
+    ,dexs.project_contract_address
+    ,dexs.tx_hash
+    ,tx.from AS tx_from
+    ,tx.to AS tx_to
+    ,dexs.trace_address
+    ,dexs.evt_index
+FROM dexs
+INNER JOIN {{ source('ethereum', 'transactions') }} tx
+    ON dexs.tx_hash = tx.hash
+    {% if not is_incremental() %}
+    AND tx.block_time >= '{{project_start_date}}'
+    {% endif %}
+    {% if is_incremental() %}
+    AND tx.block_time = date_trunc("day", now() - interval '1 week')
+    {% endif %}
+LEFT JOIN {{ ref('tokens_erc20') }} erc20a
+    ON erc20a.contract_address = dexs.token_bought_address
+    AND erc20a.blockchain = 'ethereum'
```

Comment:
missing the blockchain filter on the two joins to erc20 tokens was causing the dupes